network/member.go

```diff
@@ -0,0 +1,415 @@
+		}
+
+		err = m.handleNotification(resp)
+		if err != nil {
+			log.Println(err)
+			continue
+		}
+	}
+}
+
+// handle notification from manager
+func (m *Member) handleNotification(noti *managerToMember) error {
+	if m.opts.Verbose {
+		log.Printf("nConnect member got notification: %+v", noti)
+	}
+
+	switch noti.MsgType {
+	case NOTI_AUTHORIZED:
+		if len(noti.NodeInfo) > 0 {
+			m.networkData.NodeInfo = noti.NodeInfo[0]
+			m.saveMemberData()
+			SetNetworkIp(m.opts, m.networkData.NodeInfo.IP, m.networkData.NetworkInfo.Gateway, m.networkData.NodeInfo.Netmask, m.networkData.NetworkInfo.DNS)
+			m.GetNodeICanAccess()
+		}
+
+	case NOTI_NEW_MEMBER:
+		m.GetNodeICanAccess()
+	case NOTI_SET_ACCEPT:
+		m.GetNodeIAccept()
+	case NOTI_MEMBER_ONLINE:
+		if m.CbNodeICanAccessUpdated != nil {
+			m.CbNodeICanAccessUpdated()
+		}
+
+	default:
+		return fmt.Errorf("nConnect member got unknown notification type: %v", noti.MsgType)
+	}
+
+	return nil
+}
+
+func (m *Member) JoinNetwork() error {
+	msg := memberToManager{MsgType: JOIN_NETWORK, Name: m.opts.NodeName}
+	resp, err := SendMsg(m.mc, m.opts.ManagerAddress, &msg, true)
+	if err != nil {
+		return err
+	}
+	if resp.Err == errWaitForAuth {
+		m.networkData.NetworkInfo = resp.NetworkInfo
+		m.saveMemberData()
+		return nil
+	} else if resp.Err == errNameExist {
+		log.Printf("Node name %v already exists in this network, please use another name", m.opts.NodeName)
+		return errNameExist
+	}
+
+	if resp.Err != nil {
+		return resp.Err
+	}
+
+	if len(resp.NodeInfo) > 0 {
+		m.networkData.NodeInfo = resp.NodeInfo[0]
+		m.networkData.NetworkInfo = resp.NetworkInfo
+		m.saveMemberData()
+		if m.networkData.NodeInfo.IP != "" {
+			m.joinedNetwork = true
+			SetNetworkIp(m.opts, m.networkData.NodeInfo.IP, m.networkData.NetworkInfo.Gateway, m.networkData.NodeInfo.Netmask, m.networkData.NetworkInfo.DNS)
+		}
+	}
+
+	return nil
+}
+
+func (m *Member) LeaveNetwork() error {
+	msg := memberToManager{MsgType: LEAVE_NETWORK, Name: m.opts.NodeName}
+	resp, err := SendMsg(m.mc, m.opts.ManagerAddress, &msg, true)
+	if err != nil {
+		return err
+	}
+	if resp.Err != nil {
+		return resp.Err
+	}
+
+	m.networkData = memberNetworkData{}
+	return m.saveMemberData()
+}
+
+func (m *Member) GetMyInfo() (*nodeInfo, error) {
+	msg := memberToManager{MsgType: GET_MY_INFO}
+
+	resp, err := SendMsg(m.mc, m.opts.ManagerAddress, &msg, true)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Err != nil {
+		return nil, resp.Err
+	}
+
+	if len(resp.NodeInfo) > 0 {
+		m.networkData.NodeInfo = resp.NodeInfo[0]
+		if err = m.saveMemberData(); err != nil {
+			log.Printf("Failed to save member data: %v", err)
+		}
+		if err = SetNetworkIp(m.opts, m.networkData.NodeInfo.IP, m.networkData.NetworkInfo.Gateway, m.networkData.NodeInfo.Netmask,
+			m.networkData.NetworkInfo.DNS); err != nil {
+			log.Printf("Failed to set network ip: %v", err)
+		}
+	}
+
+	return m.networkData.NodeInfo, err
+}
+
+func (m *Member) UpdateName(name string) error {

```

Comment:
Should manager also send message to members on update event? (update name, update ip, etc)